extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -16,45 +16,175 @@
+            serialize(out, datumWriter, data);
+            return out.toByteArray();
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static <T> void serialize(ByteArrayOutputStream out, DatumWriter<T> datumWriter, T data) throws IOException {
+        BinaryEncoder encoder = EncoderFactory.get().directBinaryEncoder(out, null);
+        datumWriter.write(data, encoder);
+        encoder.flush();
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data) {
+        return deserialize(datumReader, data, 0, data.length);
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data, int offset, int length) {
+        try {
+            BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(data, offset, length, null);
+            return datumReader.read(null, decoder);
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static class GenericRecordHook implements SerializerHook<GenericRecord>, HazelcastInstanceAware {
+        private final Map<Schema, AvroSchema> replicatedSchemas = new HashMap<>();
+        private SchemaService schemaService;
 
+        @Override
+        public Class<GenericRecord> getSerializationType() {
+            return GenericRecord.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new ByteArraySerializer<GenericRecord>() {
+                @Override
+                public int getTypeId() {
+                    return SerializerHookConstants.AVRO_GENERIC_RECORD;
+                }
+
+                @Override
+                public byte[] write(GenericRecord record) {
+                    if (record instanceof LazyImmutableRecord) {
+                        return ((LazyImmutableRecord) record).serializedRecord;
+                    }
+                    try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+                        AvroSchema schema = replicatedSchemas.get(record.getSchema());
+                        if (schema == null) {
+                            schema = new AvroSchema(record.getSchema());
+                            schemaService.put(schema);
+                            replicatedSchemas.put(record.getSchema(), schema);
+                        }
+                        out.write(ByteBuffer.allocate(SCHEMA_ID_SIZE).putLong(schema.getSchemaId()).array());
+                        serialize(out, new GenericDatumWriter<>(record.getSchema()), record);
+                        return out.toByteArray();
+                    } catch (IOException e) {
+                        throw rethrow(e);
+                    }
+                }
+
+                @Override
+                public GenericRecord read(byte[] buffer) {
+                    long schemaId = ByteBuffer.wrap(buffer).getLong();
+                    Schema schema = ((AvroSchema) schemaService.get(schemaId)).getSchema();
```

Comment:
I wonder if we should fail more gracefully (and skip the message?) if the schema is unknown or non-Avro? Think eg. of copying jet snapshot to different cluster.